File: src/Recursion/BackTracking/NQueens.java
package Recursion.BackTracking;

public class NQueens {
    public static void main(String[] args) {
        int n= 4;
        boolean[][] b = new boolean[n][n];
        System.out.println(findNQueens(b,0));
    }
    static int findNQueens(boolean[][] board,int row){
        if(row==board.length){
            display(board);
            System.out.println();
            return 1;
        }
        int count = 0;
        for(int col=0;col<board[row].length;col++){
            if(isSafe(board,row,col)){
                board[row][col] = true;
                count+=findNQueens(board,row+1);
                board[row][col] = false;
            }
        }
        return count;
    }

    private static boolean isSafe(boolean[][] b,int row,int col) {
        for (int i = 0; i < row; i++) {
            if(b[i][col]){
                return false;
            }
        }
        int maxLeftDiag = Math.min(row,col);
        for (int i = 1; i <= maxLeftDiag; i++) {
            if(b[row-i][col-i]){
                return false;
            }
        }
        int maxRightDiag = Math.min(row,b[row].length-col-1);
        for (int i = 1; i <= maxRightDiag; i++) {
            if(b[row-i][col+i]){
                return false;
            }
        }
        return true;
    }

    private static void display(boolean[][] board) {
        for(boolean[] row:board){
            for(boolean elem:row){
                if(elem){
                    System.out.print("Q ");
                }
                else{
                    System.out.print("X ");
                }
            }
            System.out.println();
        }
    }
}

File: POTD/NQueens.py
# POTD 30-01-2025
# https://www.geeksforgeeks.org/batch/gfg-160-problems/track/recursion-and-backtracking-gfg-160/problem/n-queen-problem0315

class Solution:
    def nQueen(self, n):
        # code here
        board = [['.' for i in range(n)] for i in range(n)]
        return getNQueens(board,0,0)
def getNQueens(board,row,col):
    if row==len(board):
        l = []
        for arr in board:
            for j in range(len(arr)):
                if arr[j]=="Q":
                    l.append(j+1)
        return [l]
    l = []
    for i in range(len(board)):
        if(isSafe(board,row,i)):
            board[row][i] = 'Q'
            l.extend(getNQueens(board,row+1,0))
            board[row][i] = '.'
    return l
def isSafe(board,row,col):
    for i in range(row):
        if board[i][col] == "Q":
            return False
    
    # left diagonal
    left = min(col,row)
    for i in range(1,left+1):
        if(board[row-i][col-i]=="Q"):
            return False
    
    # right diagonal
    right = min(row,len(board)-col-1)
    for i in range(1,right+1):
        if(board[row-i][col+i]=="Q"):
            return False
    return True